src/cwe_checker_lib/src/abstract_domain/character_inclusion.rs

```diff
@@ -0,0 +1,293 @@
+//!     e.g. ({H,e,l,o,',',' '}, {H,e,l,o,',',' '}) v ({W,o,r,l,d}, {W,o,r,l,d}) => ({l,o}, {H,e,l,o,',',' ',W,o,r,d})
+
+use std::{collections::BTreeSet, fmt};
+
+use crate::prelude::*;
+
+use super::{AbstractDomain, DomainInsertion, HasTop};
+
+/// The `CharacterInclusionDomain` is a abstract domain describing the characters a string certainly has.
+/// and the characters a string may have.
+///
+/// The value comprises of a set of certainly contained characters and a set of possibly contained characters
+/// while the *Top* value does not get any data. However, the *Top* value stands for an empty set of certainly
+/// contained characters and the whole alphabet of allowed characters for the possibly contained characters.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub enum CharacterInclusionDomain {
+    /// The *Top* value stands for an empty set of certainly contained characters and
+    /// the whole alphabet of allowed characters for the possibly contained characters.
+    Top,
+    /// The set of certainly contained characters and a set of possibly contained characters
+    Value((CharacterSet, CharacterSet)),
+}
+
+impl CharacterInclusionDomain {
+    /// Unwraps the values from the Character Inclusion Domain
+    pub fn unwrap_value(&self) -> (CharacterSet, CharacterSet) {
+        match self {
+            CharacterInclusionDomain::Value(value) => value.clone(),
+            _ => panic!("Unexpected Character Inclusion type."),
+        }
+    }
+}
+
+impl DomainInsertion for CharacterInclusionDomain {
+    /// Append string domain as part of a concatenation. (different to merge)
+    fn append_string_domain(&self, string_domain: &Self) -> CharacterInclusionDomain {
+        match self {
+            CharacterInclusionDomain::Value((self_certain, self_possible)) => match string_domain {
+                CharacterInclusionDomain::Value((other_certain, other_possible)) => {
+                    CharacterInclusionDomain::Value((
+                        self_certain.union(other_certain.clone()),
+                        self_possible.union(other_possible.clone()),
+                    ))
+                }
+                CharacterInclusionDomain::Top => {
+                    CharacterInclusionDomain::Value((self_certain.clone(), CharacterSet::Top))
+                }
+            },
+            CharacterInclusionDomain::Top => match string_domain {
+                CharacterInclusionDomain::Value((other_certain, _)) => {
+                    CharacterInclusionDomain::Value((other_certain.clone(), CharacterSet::Top))
+                }
+                CharacterInclusionDomain::Top => CharacterInclusionDomain::Top,
+            },
+        }
+    }
+
+    /// Create a string domain that approximates float values.
+    fn create_float_value_domain() -> Self {
+        let float_character_set: BTreeSet<char> = vec![
+            '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '-', 'a', 'i', 'n', 'f',
+        ]
```

Comment:
I would have expected this set to also contain `e` or `E` for exponent. Or does standard C float formatting not print exponents?